src/Compilers/CSharp/Portable/CSharpParseOptions.cs

```diff
@@ -186,19 +187,46 @@ internal ImmutableArray<ImmutableArray<string>> InterceptorsPreviewNamespaces
                 }
 
                 // e.g. [["System", "Threading"], ["System", "Collections"]]
-                ImmutableArray<ImmutableArray<string>> previewNamespaces;
-                if (Features.TryGetValue("InterceptorsPreviewNamespaces", out var namespaces))
+                ImmutableArray<ImmutableArray<string>> previewNamespaces = Features.TryGetValue("InterceptorsPreviewNamespaces", out var namespaces) && namespaces.Length > 0
+                    ? makeNamespaces(namespaces)
+                    : ImmutableArray<ImmutableArray<string>>.Empty;
+
+                ImmutableInterlocked.InterlockedInitialize(ref _interceptorsPreviewNamespaces, previewNamespaces);
+                return previewNamespaces;
+
+                static ImmutableArray<ImmutableArray<string>> makeNamespaces(string namespaces)
                 {
-                    previewNamespaces = namespaces
-                        .Split(';')
-                        .SelectAsArray(segment => segment.Split('.').ToImmutableArray());
+                    var builder = ArrayBuilder<ImmutableArray<string>>.GetInstance();
+                    var singleNamespaceBuilder = ArrayBuilder<string>.GetInstance();
+                    int currentIndex = 0;
+                    while (currentIndex < namespaces.Length && namespaces.IndexOf(';', currentIndex) is not -1 and var semicolonIndex)
+                    {
+                        addSingleNamespaceParts(builder, singleNamespaceBuilder, namespaces.AsSpan(currentIndex, semicolonIndex - currentIndex));
+                        currentIndex = semicolonIndex + 1;
+                    }
+
+                    addSingleNamespaceParts(builder, singleNamespaceBuilder, namespaces.AsSpan(currentIndex));
+                    singleNamespaceBuilder.Free();
+                    return builder.ToImmutableAndFree();
                 }
-                else
+
+                static void addSingleNamespaceParts(ArrayBuilder<ImmutableArray<string>> namespacesBuilder, ArrayBuilder<string> singleNamespaceBuilder, ReadOnlySpan<char> @namespace)
                 {
-                    previewNamespaces = ImmutableArray<ImmutableArray<string>>.Empty;
+                    int currentIndex = 0;
+                    while (currentIndex < @namespace.Length && @namespace.IndexOf('.', currentIndex) is not -1 and var dotIndex)
+                    {
+                        singleNamespaceBuilder.Add(@namespace.Slice(currentIndex, dotIndex - currentIndex).ToString());
```

Comment:
Not related to this PR: should we handle spaces or other kind of whitespaces? It's probably possible to inject those by putting the whole list into quotes, no?

Reply:
This comes from an msbuild property so I imagine something as simple as `<Features>InterceptorsPreviewNamespaces=NS1. NS2</Features>` would do it. There are a couple "sad case" tests with leading and trailing spaces in `FeatureFlag_Granular_02`. I'm not sure if there's any change needed in this feature parsing code to handle whitespace better.

Reply:
I would probably have handled starting/trailing whitespaces, but I guess it's okay to wait until we hear some complaints.